src/Eshopworld.Messaging/Messenger.cs

```diff
@@ -0,0 +1,164 @@
+        /// <param name="subscriptionId">The subscription ID where the service bus namespace lives.</param>
+        public Messenger([NotNull]string connectionString, [NotNull]string subscriptionId)
+        {
+            ConnectionString = connectionString;
+            SubscriptionId = subscriptionId;
+        }
+
+        /// <inheritdoc />
+        public async Task Send<T>(T message)
+            where T : class
+        {
+            if (!QueueAdapters.ContainsKey(typeof(T))) // double check, to avoid locking it here to keep it thread safe
+            {
+                SetupMessageType<T>(10);
+            }
+
+            await ((MessageQueueAdapter<T>)QueueAdapters[typeof(T)]).Send(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void Receive<T>(Action<T> callback, int batchSize = 10)
+            where T : class
+        {
+            lock (Gate)
+            {
+                if (MessageSubs.ContainsKey(typeof(T)))
+                {
+                    throw new InvalidOperationException("You already added a callback to this message type. Only one callback per type is supported.");
+                }
+
+                SetupMessageType<T>(batchSize).StartReading();
+
+                MessageSubs.Add(
+                    typeof(T),
+                    MessagesIn.OfType<T>().Subscribe(callback));
+            }
+        }
+
+        /// <inheritdoc />
+        public void CancelReceive<T>()
+            where T : class
+        {
+            lock (Gate)
+            {
+                if (!QueueAdapters.ContainsKey(typeof(T)))
+                {
+                    throw new InvalidOperationException($"Messages of type {typeof(T).FullName} haven't been setup properly yet");
+                }
+
+                ((MessageQueueAdapter<T>) QueueAdapters[typeof(T)]).StopReading();
+
+                MessageSubs[typeof(T)].Dispose();
+                MessageSubs.Remove(typeof(T));
+            }
+        }
+
+        /// <inheritdoc />
+        public IObservable<T> GetObservable<T>(int batchSize = 10) where T : class
+        {
+            SetupMessageType<T>(batchSize).StartReading();
+
+            return MessagesIn.OfType<T>().AsObservable();
+        }
+
+        /// <inheritdoc />
+        public async Task Lock<T>(T message) where T : class
+        {
+            var adapter = (MessageQueueAdapter<T>) QueueAdapters[typeof(T)];
+            await adapter.Lock(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public async Task Complete<T>(T message) where T : class
+        {
+            var adapter = (MessageQueueAdapter<T>) QueueAdapters[typeof(T)];
+            await adapter.Complete(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public async Task Abandon<T>(T message) where T : class
+        {
+            var adapter = (MessageQueueAdapter<T>) QueueAdapters[typeof(T)];
```

Comment:
if the dev calls this without subscribing etc.. these would be null I think, this also applies to other methods here potentially I think